docs/design/plan-modification.md

```diff
@@ -0,0 +1,543 @@
+
+#### Tradeoffs
+
+Similarly to option 3, the plan modifier functions are easily unit testable, and the `PlanModifier` field easily discoverable on the `schema.Attribute` struct. The code is reasonably Go-native.
+
+This solution aims to improve on the compatibility of option 3 by ensuring that no further fields need be added to `schema.Attribute` in future for the purposes of plan modification.
+
+### 4a. `attr.ValueAs()`
+
+This option is to be considered an extension to option 4.
+
+Plan modifier functions (`AttributePlanModifier.Modify()`), and helper functions such as `RequiresReplaceIf`, are supplied `attr.Value` arguments from which the provider code must derive attribute values. In the above example, this is done with type assertions:
+
+```go
+RequiresReplaceIf(func(ctx context.Context, state, config attr.Value) (bool, error) {
+  stateVal := state.(types.Number)
+  configVal := config.(types.Number)
+  // ...
+}
+```
+
+Avoiding such type assertions is one of the design goals of the new framework.
+
+Solution 4 should therefore be considered in combination with a way of avoiding such type assertions in provider code. One such proposal, which will be included in a separate design doc, is an `attr.ValueAs()` function:
+
+```go
+func ValueAs(ctx context.Context, val attr.Value, target interface{}) error {}
+```
+
+Similar to functions such as `state.Get(ctx, val, target)`, this function could use reflection to populate `target` with the value in `val`, returning an error if `target` is not of a compatible type. The provider code could then be:
+
+
+```go
+RequiresReplaceIf(func(ctx context.Context, state, config attr.Value) (bool, error) {
+  var stateInt int
+  var configInt int
+  err: = attr.ValueAs(ctx, state, stateInt)
+  if err != nil {
+  	return false, err
+  }
+    err: = attr.ValueAs(ctx, config, configInt)
+  if err != nil {
+  	return false, err
+  }
+  // ...
+}
+```
+
+### 5. `attr.TypeWithModifyPlan`
+
+A `ModifyPlan()` or `RequiresReplaceIf()` function could be added to an extension of the `attr.Type` interface:
+
+```go
+type TypeWithModifyPlan interface {
+  Type
+  
+  ModifyPlan(context.Context, ModifyAttributeTypePlanRequest, *ModifyAttributeTypePlanResponse) bool
+```
+
+This would allow bundling reusable `ModifyPlan` behaviour up with a custom type's validation and other behaviours. This could be useful, for example, in a custom timestamp type to squash semantically meaningless diffs, so provider developers do not have to specify the attribute plan modifier wherever the attribute appears in a schema.
+
+Without knowing how custom types will be used by provider developers, this option seems premature, and makes less sense than bundling validation functions with custom types.
```

Comment:
I disagree, I think it makes as much sense, but also am fine with adding it later given we've called out the design and use case here and acknowledged that the only reason we're not doing it is we're waiting for a use case to present itself, and not because we think the design is flawed or problematic or something.

Reply:
I'm 👍 (which I think we all are!) for having custom type design be its own discussion and I don't think this design's recommendations are particularly preventing those efforts.